opencog/spacetime/TimeSpaceAtomMapsAPI/TimeSpaceAtom.cc

```diff
@@ -0,0 +1,316 @@
+ *     * Neither the name of the University of Freiburg nor the names of its
+ *       contributors may be used to endorse or promote products derived from
+ *       this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+//TimeSpaceAtom.cpp
+#include "TimeSpaceAtom.h"
+//#include "octomap/OcTreeKey.h"
+#include <assert.h>
+
+TimeSpaceAtom::TimeSpaceAtom(unsigned int num_time_units,
+                             vector<double>map_res_meters):
+    time_circle(num_time_units), created_once(false)
+{
+    map_count = map_res_meters.size(); //if zero throw exception
+    int i = 0;
+    for_each(map_res_meters.begin(), map_res_meters.end(), [&](double resolution) {
+        i++;
+        map_res[i] = resolution;
+    });
+}
+
+unsigned int
+TimeSpaceAtom::GetMapCount()
+{
+    return map_count;
+}
+
+bool
+TimeSpaceAtom::GetMapResolution(const int handle, double& res)
+{
+    auto it = map_res.find(handle);
+    if (it == map_res.end()) return false;
+    res = it->second;
+    return true;
+}
+
+bool
+TimeSpaceAtom::GetCurrentTimeRange(time_pt& time_p, duration_c& duration)
+{
+    if (time_circle.size() < 1) return false;
+    time_p = curr_time;
+    duration = curr_duration;
+    return true;
+}
+
+bool
+TimeSpaceAtom::CreateNewTimeUnit(const time_pt time_p,
+                                 const duration_c duration)
+{
+    //ideally greater than check should be done
+    if (created_once) {
+        if (IsTimePointInRange(time_p, curr_time, curr_duration) ||
+                IsTimePointInRange(time_p + duration, curr_time, curr_duration)) {
+            return false;
+        }
+    }
+
+    time_unit temp(time_p, duration);
+    time_circle.push_back(temp);
+    int i = time_circle.capacity() - 1;
+    if (time_circle.size() < time_circle.capacity()) i = time_circle.size() - 1;
+    for_each( map_res.begin(), map_res.end(), [&](pair<int, double> handle) {
+        time_circle[i].map_tree[handle.first].setResolution(handle.second);
+    }
+    );
+
+    curr_time = time_p;
+    curr_duration = duration;
+    created_once = true;
+    return true;
+}
+
+bool
+TimeSpaceAtom::PutAtomAtCurrentTime(const int map_handle,
```

Comment:
Could you explain to me this choice, why map_handle are int?

How is opencog gonna represent a map?